RT10xx Power Management: Enable RT1060 Soft Off Mode and Fixup PM API

NickolasLapp opened the pull request:
This PR adds a soft off mode to the RT10xx Power Management API, and adds sample code demonstrating the usage of this power mode. 

Signed-off-by: Nickolas Lapp <nickolaslapp@gmail.com>

danieldegrasse commented on soc/arm/nxp_imx/rt/power_rt10xx.c:
```diff
@@ -76,6 +76,13 @@ static void lpm_set_sleep_mode_config(clock_mode_t mode)
 	GPC_DisableIRQ(GPC, GPR_IRQ_IRQn);
 }
 
+static void lpm_enter_soft_off_mode(void)
```
Hi @NickolasLapp, thank you for the contribution here. I think enabling SNVS soft off mode for the RT series would be valuable, but I am concerned about doing so without a simple sample application (something like `samples/boards/nrf/system_off`.

Without a sample application, testing this contribution becomes difficult, and the value to the community is lower, since users don't have a demonstration of how to enter SNVS mode on their RT board without first writing their own Zephyr application. Beyond this, the GPC supports a variety of wakeup sources, and adding a sample application would ensure that the RTC was always set up with an alarm, as well as help users understand the relationship between the GPC api call here and the setup required before entering low power mode.

NickolasLapp commented:
@danieldegrasse I added a sample of using the force-power off (modeled after the example you provided). I also added support for the RT1060EVK-B since that's the board I have available to test. Please take a look and let me know what you think.

danieldegrasse reviewed (commented):
@NickolasLapp Thank you for creating the demo app- tested on an EVKB and it works for me. Could you pull the changes enabling the EVKB into another commit, and update the title of this PR accordingly? Beyond that and the CI errors, this PR LGTM to me.

NickolasLapp committed:
i.MX NXP Boards: Add support for RT1060_EVKB

This PR adds support for the RT1060_EVKB as a variant of the RT1060 EVK.
Blinky app tested locally on RT1060_EVKB.

Signed-off-by: Nickolas Lapp <nickolaslapp@gmail.com>

NickolasLapp committed:
RT10xx Power Management: Enable RT1060 Soft Off Mode and Fixup PM API

This PR adds a soft off mode to the RT10xx Power Management API.
Additionally, it corrects the PM API function in rt10xx_power.c to
use the correct function prototype to be properly overridden.

Signed-off-by: Nickolas Lapp <nickolaslapp@gmail.com>

NickolasLapp commented:
> @NickolasLapp Thank you for creating the demo app- tested on an EVKB and it works for me. Could you pull the changes enabling the EVKB into another commit, and update the title of this PR accordingly? Beyond that and the CI errors, this PR LGTM to me.

Hi @danieldegrasse, I split the PR into two commits (one for adding the EVKb). I also fixed some (hopefully most) of the CI issues. Will see what's remaining when everything runs through here.

danieldegrasse commented on tests/drivers/uart/uart_async_api/src/test_uart.h:
```diff
@@ -57,6 +57,7 @@
 #define UART_DEVICE_NAME DT_LABEL(DT_NODELABEL(lpuart2))
 #elif defined(CONFIG_BOARD_MIMXRT1050_EVK) || \
 	defined(CONFIG_BOARD_MIMXRT1060_EVK) || \
+	defined(CONFIG_BOARD_MIMXRT1060_EVKB) || \
```
Non blocking suggestion- I believe you need to add a board specific overlay for `mimxrt1060_evkb` if you want to enable LPUART3 for this test, and relocate sram into DTCM to avoid caching issues.

NickolasLapp replied:
Added an overlay for this

danieldegrasse reviewed (dismissed):
One minor nitpick about the UART async API test, other than that and the CI failures this PR LGTM to me

danieldegrasse commented on samples/boards/mimxrt1060_evk/system_off/src/main.c:
```diff
@@ -0,0 +1,95 @@
+#include <stdio.h>
```
this file needs a copyright to pass CI

danieldegrasse commented on samples/boards/mimxrt1060_evk/system_off/Kconfig:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2022, NXP
```
copyright holder

danieldegrasse commented on boards/arm/mimxrt1060_evk/mimxrt1060_evkb_defconfig:
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) 2022, NXP
+#
+# SPDX-License-Identifier: Apache-2.0
```
Ditto on copyright holder

danieldegrasse commented on boards/arm/mimxrt1060_evk/mimxrt1060_evkb.yaml:
```diff
@@ -0,0 +1,32 @@
+#
+# Copyright (c) 2022, NXP
```
Copyright holder

danieldegrasse commented on boards/arm/mimxrt1060_evk/mimxrt1060_evkb.dts:
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2022, NXP
```
Assuming you want to change this copyright

danieldegrasse commented on tests/drivers/uart/uart_async_api/boards/mimxrt1060_evkb.overlay:
```diff
@@ -0,0 +1,12 @@
+
+&lpuart3 {
+	status = "okay";
+	current-speed = <115200>;
```
Not blocking/strictly required, but if you add the `nxp,loopback` property here the test can be run without physically connecting the TX/RX of the LPUART pins

danieldegrasse commented on west.yml:
```diff
@@ -101,7 +101,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: 46be8173d2453b16c617b555f96c59d62920a290
+      revision: d1330d29486ba8d86ea10e9eedae8ce5b39f5527
```
Due to the order that changes merged in NXP's HAL, this PR is dependent on #44961. Essentially, updating the SHA of NXP's HAL to `d1330d29486ba8d86ea10e9eedae8ce5b39f5527` before #44961 merges would cause build failures on some of NXP's platforms, that #44961 contains the updates required to fix. I'm going to mark this PR as DNM rather than NAKing it, since there is no issue with the PR itself.

dleach02 replied:
@NickolasLapp I think we need to just remove the west.yml changes from this because the version in Zephyr now has a SHA that is past this SHA so we can't go backwards

danieldegrasse reviewed (dismissed):
See my comment on `west.yml`- PR LGTM but I am adding DNM to it until #44961 merges

NickolasLapp commented:
@danieldegrasse I think #44961 has landed if you don't mind removing the DNM label here. I also fixed the nxp loopback property for the uart overlay for the evk-b

dleach02 reviewed (dismissed):
@NickolasLapp please resolve the merge conflict. You really just need to remove the west.yml update from this PR since your original HAL update already merged.

NickolasLapp commented:
@dleach02 @danieldegrasse west.yml conflict resolved, sorry about the delay--just got back from PTO.

dleach02 commented:
@NickolasLapp, you comment that this PR:

> Additionally, it corrects the PM API function in rt10xx_power.c to use the correct function prototype to be properly overridden.

Please add a github issue for the problem being fixed. Then add a "Fixes #<nnn>" to the description above to link this PR and establish traceability. This release cycle is in feature freeze mode.

NickolasLapp commented:
> @NickolasLapp, you comment that this PR:
> 
> > Additionally, it corrects the PM API function in rt10xx_power.c to use the correct function prototype to be properly overridden.
> 
> Please add a github issue for the problem being fixed. Then add a "Fixes #" to the description above to link this PR and establish traceability. This release cycle is in feature freeze mode. Doing this will also not dismiss any positive reviews.


@dleach02 the mentioned issue was already resolved in https://github.com/zephyrproject-rtos/zephyr/commit/99be4358afcca4346acfa79fbf92286a6d934af0 (and is no longer in this PR as this PR was rebased on main). I've updated the description of this PR to indicate that it only adds support for the soft off mode for NXP parts, and adds a sample demonstrating this usage.

danieldegrasse reviewed (approved):
LGTM to me, although this likely won't be merged until after the 3.1 release since we are currently in a feature freeze state.
